### ISSUE TITLE
Accidentally using :class instead of :class_name in association gives misleading error

### ISSUE THREAD

**Issue by pixeltrix** [rails/rails#19659]
If you define an association using a custom class and then make the all-too common typo of using `:class` instead of `:class_name` then the change made in 1f006cd will give rise to a hard to track down error in [`raise_on_type_mismatch!`][1] because `is_a?` is called with a string instead of a class.

Since `:class` option is private and added for supporting HABTM associations on top of HMT then there is a question whether we should change the option name to be something else less prone to error or try to detect the error when building the association.

wdyt? @tenderlove @sgrif @rafaelfranca @carlosantoniodasilva 

[1]: https://github.com/rails/rails/blob/v4.2.1/activerecord/lib/active_record/associations/association.rb#L213-L218

**Comment by carlosantoniodasilva** [comment]
Right, I agree it might be easy for someone use `:class` instead of `:class_name`. Maybe we should use a different name like `anonymous_class` or `klass` instead?

**Comment by rafaelfranca** [comment]
Renaming seems fine.

**Comment by sgrif** [comment]
`anonymous_class` is pretty intention revealing to me. I wouldn't mind `_internal_class` or something similar, either.

**Comment by pascalbetz** [comment]
+1 on ```anonymous_class```

(someone who made this typo as well)

**Comment by meinac** [comment]
Why nobody don't think about giving meaningful error message instead of renaming the option name. class is really fit for this option name.
If I was a programmer who don't know the option name and want to assign a class to my relation I would try to use class option without looking to the documentation.

**Comment by carlosantoniodasilva** [comment]
@meinac the point is that this option is for internal use only, not for people to consume, and we'd have no way to differentiate when it's being used by the framework or by a developer on its own code (that's why they should not - consider this as private, an implementation detail).

**Comment by dimitko** [comment]
I'm not a Rails core dev obviously, but I'm just hoping in to tell you that `anonymous_class` makes no sense at all to me. Similar to what @meinac stated, as a dev who relies on Rails' convention-over-configuration general approach, I'd naively just go for the `class` option and then scratch my head for 2 mins before ending up right here in this thread.

I simply opted for using the `class_name` option everywhere. As a guy managing several projects I value the low numbers of times I ask myself "WTF was I thinking when I coded this?" several weeks/months down the line.

BTW, something like `association_class` would be much more immediately apparent.
